### PR TITLE
Implemented "Write to file" for UFunction watches in the watches tab & fixed a crash on shutdown

### LINKED ISSUE #418: [BUG - Release] Crash when exiting game with watches to be logged to file
**Branch or Release**
3.0.1

**Game and Engine Version**
Deep Rock Galactic, 4.27

**Describe the bug**
Everything running as normal, using watches on some stuff fine, with logging to file selected. But when I close the game, I get a crash dmp and none of the data that was sent to the watches on the GUI was output into any file. I thought it's meant to get sent to watches.meta.json but that file seems to only be populated when you save a watch to be auto set when you next load up the game.

**Mods directory**
No mods installed.

**To Reproduce**
It's a bit odd, I cannot reliably repro the crash part issue with the same setup steps as below, however my watched data (except saving watch for auto-load) are never saved to any file.

1. Launch game
2. Add watch to `/Script/GameplayCameras.MatineeCameraShake:ReceiveIsFinished`
3. Load into mission (so there's a few entries in there)
4. Exit game
5. Crash (sometimes)

**Expected behavior**
Game to not crash, and the write to file setting on a watch to actually write the data to a file.

**Screenshots, UE4SS Log, and .dmp file**
[UE4SS.log](https://github.com/UE4SS-RE/RE-UE4SS/files/14454514/UE4SS.log)

[crash_2024_02_29_22_57_44.dmp](https://cdn.discordapp.com/attachments/1109192354595876944/1212901970092888064/crash_2024_02_29_22_57_44.dmp?ex=65f385fc&is=65e110fc&hm=329e039933dee73c3b5d56003f202f2defe9f341ae0c23a54bde4d2749ef5493&)

**Desktop:**
 - OS: Win10

### PR DESCRIPTION
**Pull request by UE4SS**
I logged this as a fix in the changelog because even though it's technically a new feature, it's much more likely to be seen as a bug by people.

**Description**

See commit messages.

Partially fixes #418

**Type of change**

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

**How Has This Been Tested?**

I tested this by adding a watch to `CanJumpInternal` and ticking `Write to file` and proceeding to jump in the game and then confirming that `ue4ss_watch_Character.CanJumpInternal_.txt` existed in the `watches` directory.

**Checklist**

- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added the necessary description of this PR to the changelog, and I have followed the same format as other entries.
- [x] Any dependent changes have been merged and published in downstream modules.

**Screenshots (if appropriate)**

**Additional context**

Add any other context about the pull request here.

### CODE REVIEW

**Comment by UE4SS** [comment]
This PR has been updated to fully fix #418.

**Review by Foxydapirate12 — APPROVED** [review]

**Comment by Buckminsterfullerene02** [comment]
Nice. The writing the file works and cannot reproduce crash at all anymore. Have found another bug though kind of related to the watch stuff, so will make a bug report and will see if you end up just bunging it into here or not (if you find a fix).

**Comment by Buckminsterfullerene02** [comment]
It would be nice if the write to file option was also saved by the watches json.

**Comment by UE4SS** [comment]
This PR has been expanded again, do not merge until I confirm that it's ready.

**Comment by UE4SS** [comment]
> > Nice. The writing the file works and cannot reproduce crash at all anymore. Have found another bug though kind of related to the watch stuff, so will make a bug report and will see if you end up just bunging it into here or not (if you find a fix).
> 
> Nevermind, I got a crash again (with dll built from this branch), though this time no crash dmp. Can reproduce reliably. Always crashes when I start a mission and am in the loading screen, with the following watches.meta.json:
> 
> ```json
> {
>   "Watches":[
>     {
>       "AcquisitionID":"/Script/GameplayCameras.MatineeCameraShake:ReceiveIsFinished",
>       "PropertyName":" ",
>       "AcquisitionMethod":0,
>       "WatchType":1
>     },
>     {
>       "AcquisitionID":"/Script/Engine.Default__CameraModifier_CameraShake",
>       "PropertyName":"ExpiredPooledShakesMap",
>       "AcquisitionMethod":0,
>       "WatchType":0
>     },
>     {
>       "AcquisitionID":"/Script/Engine.Default__CameraModifier_CameraShake",
>       "PropertyName":"SplitScreenShakeScale",
>       "AcquisitionMethod":0,
>       "WatchType":0
>     }
>   ]
> }
> ```
> 
> and the only watch set to write to file is the top one.

I'm unable to reproduce this crash.
It also appears to be a different crash than #418 since that one triggers on game exit and this one triggers when the game loads.
Can you check if this new crash happens on the main branch ? I want to know if any of my changes in this pull request is the cause.

**Comment by Buckminsterfullerene02** [comment]
Can't reproduce on main

**Comment by UE4SS** [comment]
Can you try this branch again ?
I made a temporary test implementation for the DeleteListener lock, let's see if that helps, and there's no need for you to update the Unreal submodule for this test.
This PR cannot be merged until this implementation has been removed or replaced.
Also I forgot to change the last commit message it.

**Comment by UE4SS** [comment]
Nevermind, as I suspected, the DeleteListener lock is not enough either.
It GCs everything, and then locks the DeleteListener lock so by the time we lock it it's already too late.
This can be seen in the UE source code: `/Engine/Source/Runtime/CoreUObject/Private/UObject/Obj.cpp` in the function `StaticExit`.
I think we need the GC lock to fix this crash.

**Comment by UE4SS** [comment]
> Can't reproduce on main

I have no idea what the problem is.
This PR can't move forward until this is figured out since the crash appears to be introduced by this PR somehow, and it makes no sense that I can't reproduce it.
I start the game, tick the "Write to file" checkbox on the first item, and then I walk to the mission selector and select the first mission and then I climb the ladder to the mission dispatch area and enter the loading screen and it loads the mission without crashing.

**Comment by UE4SS** [comment]
> Have found another bug though kind of related to the watch stuff, so will make a bug report and will see if you end up just bunging it into here or not (if you find a fix).

I don't want to bundle a fix for this together with this PR.
I also haven't looked into the problem.

**Comment by UE4SS** [comment]
@Buckminsterfullerene02 Can you attach VS to the game before it crashes and catch the crash ?
There's a feature in VS that lets you generate dump files so if you can catch it in VS, I can potentially still look at a dump file and figure this problem out.

**Comment by Buckminsterfullerene02** [comment]
> @Buckminsterfullerene02 Can you attach VS to the game before it crashes and catch the crash ? There's a feature in VS that lets you generate dump files so if you can catch it in VS, I can potentially still look at a dump file and figure this problem out.

I can't replicate my original crash... it happened twice a row back then. But I tried thrice with the exact same settings with the same dll as then and it's been fine. Since then I've rebooted my PC so maybe it's just down to "windoze moment".

**Comment by UE4SS** [comment]
> > @Buckminsterfullerene02 Can you attach VS to the game before it crashes and catch the crash ? There's a feature in VS that lets you generate dump files so if you can catch it in VS, I can potentially still look at a dump file and figure this problem out.
> 
> I can't replicate my original crash... it happened twice a row back then. But I tried thrice with the exact same settings with the same dll as then and it's been fine. Since then I've rebooted my PC so maybe it's just down to "windoze moment".

Alright.
I'll pull this out of draft and let the rest of the maintainers decide whether to merge this.
There is still the on-exit crash but I'm fairly sure it can't be solved without the GC lock.

**Review by narknon — APPROVED** [review]

**Pull request merged**
